### DISCUSSION TITLE
arduino: cannot detect usb boards

### DISCUSSION THREAD

**Discussion by esp13** [netblue30/firejail#5562] [Q&A]
Hello,

I'm trying to use lastest arduino IDE version with FireJail.
Now Arduino IDE is available as AppImage and it seem to be the suggested installation method (https://docs.arduino.cc/software/ide-v2/tutorials/getting-started/ide-v2-downloading-and-installing)

So I wanted to take this fresh new installation opportunity to setup it with FireJail.

I tired on Linux Mint 20.1 Ulyssa with FireJail version 0.9.62 and arduino-ide_2.0.3_Linux_64bit.AppImage
(I gived some tries on LMDE 4 (debian buster) with FireJail version 0.9.64 (backports) too)

I tried the default arduino.profile but quickly get some troubles to be able to detect my arduino/ESP boards on ttyUSB0 and on wifi OTA.
By disabling one by one the parameters was able to detect the boards.
I commented this lines:
```
#include disable-exec.inc
#include disable-programs.inc
#protocol unix,inet,inet6
```

With my custom lines my FireJail profile is now this one:
```
#donne un nom a la sandbox pour pouvoir tuer le processus firejail --shutdown=arduino
name arduino
#pour pouvoir acceder au repertoire sur /home meme si pas dans le sous-rep de l'user courant
allusers
#emplacement du repertoire user (faux home) specifique arduino
private /home/myhomefolder/FJ/AHome/
#on interdit les autres répertoires de /home sauf celui de l'utilisateur courant qui sera substitué par la ligne precedente
noblacklist ~
blacklist /home/*
#YBT pour bloquer les autres disques
disable-mnt
#YBT pour les tests avec nemo : Pour que le nemo lancé dans firejail ne puisse pas communiquer avec le nemo hors firejail
nodbus


# Firejail profile for arduino
# Description: AVR development board IDE and built-in libraries
# This file is overwritten after every install/update
# Persistent local customizations
include arduino.local
# Persistent global definitions
include globals.local

#noblacklist ${HOME}/.arduino15
#noblacklist ${HOME}/Arduino
#noblacklist ${DOCUMENTS}

# Allow java (blacklisted by disable-devel.inc)
#include allow-java.inc

include disable-common.inc
include disable-devel.inc
#include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
#include disable-programs.inc
include disable-xdg.inc

caps.drop all
netfilter
no3d
nodvd
# nogroups
nonewprivs
noroot
nosound
notv
novideo
#protocol unix,inet,inet6
seccomp
shell none

private-cache
private-tmp
```

But I still can't program my boards on ttyUSB0 because I get a `avrdude: ser_open(): can't open device "/dev/ttyUSB0": Permission denied` error.

Outside of FireJail this error is usually solved by this group modifications:
```
sudo usermod -a -G tty $USER
sudo usermod -a -G dialout $USER
```

But with FireJail this isn't enought.

What is the good way to get ttyUSBx permission inside FireJail?

I tried:
- `ignore nogroups` added at the end of my custom profile
- The arduino.profile from the lastest FireJail version https://github.com/netblue30/firejail/blob/master/etc/profile-a-l/arduino.profile

I founded this:
- https://github.com/netblue30/firejail/issues/5292
- https://github.com/kmk3/firejail/commit/b6a21daad93eb1e41c9afb351f027121cacffad8
- https://github.com/netblue30/firejail/issues/2042
- https://github.com/netblue30/firejail/issues/4324

But don't have enough knowledge to make it working, can you help me finding what I am missing?
(sorry for my bad English skills)

**Comment by rusty-snake** [comment]
Try to play with ignoring no groups, noroot, caps.drop, nonewprivs and seccomp.

I didn't read everything yet.

**Reply by esp13** [reply]
> Try to play with ignoring no groups, noroot, caps.drop, nonewprivs and seccomp.
> 
> I didn't read everything yet.

Thank you for your answer, I tried to comment:
- `#caps.drop all`
- `#nonewprivs`
- `#noroot`
- `#seccomp`

`# nogroups` was already commented

But I still have the `avrdude: ser_open(): can't open device "/dev/ttyUSB0": Permission denied` error.

**Reply by rusty-snake** [reply]
Then a back a step and make sure it works with `--noprofile`.

**Reply by esp13** [reply]
> Then a back a step and make sure it works with `--noprofile`.

:/ I was hopping to avoid this step because I want to keep all the files generated by the tries inside the fake home folder so I can at anytime delete all easily without polluting my real home folder.

**Reply by rusty-snake** [reply]
Or use `--noprofile --private` which shouldn't make a differences for devices.

**Reply by esp13** [reply]
> Or use `--noprofile --private` which shouldn't make a differences for devices.

I tried with all things commented inside my custom profile except :
- `name...`
- `private /home/...`
- `private-cache`
- `private-tmp`
I still get the same error.

I don't have this error on same computer with the old arduino IDE installed and launched outside FireJail

**Reply by rusty-snake** [reply]
How do you start it?

**Reply by esp13** [reply]
> How do you start it?

I have a bash script:
`cd /home/myhomefolder/FJ/AHome && firejail --appimage --profile=/home/myhomefolder/FJ/My_FireJail_arduino.profile ./arduino-ide_2.0.3_Linux_64bit.AppImage --appimage-extract-and-run --no-sandbox`

**Reply by rusty-snake** [reply]
`--appimage` implies `nogroups`.

**Reply by esp13** [reply]
> `--appimage` implies `nogroups`.

Damn!
Sorry for the trouble, now with the --appimage removed from the script and the profile still commented the error doesn't appear anymore :)

I will now try to set back each parameter commented

**Reply by esp13** [reply]
I tried one by one and I had to keep this one commented to keep the error down:
- `#noroot`

My full profile is now this one:
```
#donne un nom a la sandbox pour pouvoir tuer le processus firejail --shutdown=arduino
name arduino
#pour pouvoir acceder au repertoire sur /home meme si pas dans le sous-rep de l'user courant
allusers
#emplacement du repertoire user (faux home) specifique arduino
private /home/myhomefolder/FJ/AHome/
#on interdit les autres répertoires de /home sauf celui de l'utilisateur courant qui sera substitué par la ligne precedente
noblacklist ~
blacklist /home/*
#YBT pour bloquer les autres disques
disable-mnt
#YBT pour les tests avec nemo : Pour que le nemo lancé dans firejail ne puisse pas communiquer avec le nemo hors firejail
nodbus

# Firejail profile for arduino
# Description: AVR development board IDE and built-in libraries
# This file is overwritten after every install/update
# Persistent local customizations
include arduino.local
# Persistent global definitions
include globals.local

#noblacklist ${HOME}/.arduino15
#noblacklist ${HOME}/Arduino
#noblacklist ${DOCUMENTS}

# Allow java (blacklisted by disable-devel.inc)
#include allow-java.inc

include disable-common.inc
include disable-devel.inc
#include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
#include disable-programs.inc
include disable-xdg.inc

caps.drop all
netfilter
no3d
nodvd
# nogroups
nonewprivs
#noroot
nosound
notv
novideo
#protocol unix,inet,inet6
seccomp
shell none

private-cache
private-tmp
```

And the start script is now:
- `cd /home/myhomefolder/FJ/AHome && firejail --profile=/home/myhomefolder/FJ/My_FireJail_arduino.profile ./arduino-ide_2.0.3_Linux_64bit.AppImage --appimage-extract-and-run --no-sandbox`

Do you think this configuration is safe or it is possible to restrict more?
